### PR TITLE
feat: add publisher-path authorization rules and cross-tenant deny to gateway AuthPolicy

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Description

The gateway-level AuthPolicy only had `inference-access` rules that work for per-participant paths but don't cover [publisher paths](https://github.com/kserve/kserve/pull/5822) (`/publishers/<ns>/models/<name>/...`), which need model-level authorization against `serving.opendatahub.io/models`.

Adds `model-access-path` and `model-access-path-delegate` authorization rules for direct publisher-path traffic. Supports multi-segment model names (e.g. `facebook/opt-125m`) via `split('/models/')[1].split('/v1/')[0]` extraction. Model names containing a literal `/v1/` segment are not supported - same structural limitation as the kserve HTTPRoute template.

Adds a `deny-misrouted-model-header` rule that blocks requests where the routing identity (model header) and authorization identity (request path) diverge. Without this, a per-participant path with a valid model routing header would be authorized against the path (tenant-A) but routed via the header to a different tenant's backend (tenant-B). The deny rule uses Authorino's `patternMatching` with `predicate: "false"` for an immediate local 403 with no API server round-trip.

The model routing header name is read from the `inferenceservice-config` ConfigMap (`modelBasedRoutingHeaderName`), lowercased to match Authorino's key normalization, and validated as an HTTP token. The GatewayReconciler watches the ConfigMap's `ingress` key for changes.

BBR support (`resolvedPath` normalization for `/v1/` + header traffic) is a planned follow-up.

## How Has This Been Tested?

- Unit tests verify template rendering: 5 authorization rules, overrides
- E2E tests (13 publisher-path + 20 batch = 33 total) against a kind cluster with Kuadrant: publisher-path access, delegation, anti-spoofing, namespace collisions, cross-tenant deny, multi-segment model names
- Kind-based CI workflow triggered on auth-related file changes
- All existing batch AuthPolicy e2e tests pass

## Merge criteria:

- [ ] JIRA(s) are linked in the PR description
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

* **New Features**
  * Added configurable model-routing header support, including validation and a default header.
  * Added authorization for publisher model paths, multi-segment model names, and delegated access.
  * Added safeguards against cross-tenant model misrouting and identity spoofing.
  * Auth policies now refresh when ingress routing configuration changes.

* **Bug Fixes**
  * Improved path handling for inference, publisher, `/v1/`, and short health-check requests.
  * Added safer fallback behavior for missing or invalid routing configuration.

* **Tests**
  * Expanded unit and end-to-end coverage for authorization, routing, delegation, and edge cases.

### CODE REVIEW

**Comment by pierDipi** [comment]
@coderabbitai full review

File: internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml [hunk 1]
```diff
@@ -19,15 +25,109 @@ spec:
...
+          resolvedPath:
+            expression: >-
+              request.path.startsWith('/v1/')
+              && '{{.ModelRoutingHeader}}' in request.headers
+              && request.headers['{{.ModelRoutingHeader}}'].matches('^publishers/[^/]+/models/.+$')
+              && !(request.path == '/v1/files' || request.path.startsWith('/v1/files/')
+                || request.path == '/v1/batches' || request.path.startsWith('/v1/batches/')
+                || request.path == '/v1/models' || request.path.startsWith('/v1/models/'))
+              ? ('/' + request.headers['{{.ModelRoutingHeader}}'] + request.path)
+              : request.path
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml]: Not sure about this one. CEL split is string-based, no actual traversal. SAR rejects the literal messed up name? Do we want to tighten it and risk rejecting legitimate models? Worth double checking?

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+	// model-access-path SAR checks a nonsensical name -> 403 even with broad RBAC
+	// (resourceName doesn't match any grant).
+	headers := map[string]string{
+		"x-gateway-model-name": fmt.Sprintf("publishers/%s/models/echo-server/../../admin", f.ns),
+	}
+	resp, _ := batchEnv.gatewayGet(t, "/v1/chat/completions", f.noAccessToken, headers)
+	if resp.StatusCode != http.StatusForbidden {
+		t.Fatalf("expected 403 (SAR rejects nonsensical model name), got %d", resp.StatusCode)
+	}
+}
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go]: Addressed in previous push?

File: internal/controller/serving/llm/gateway_controller.go [hunk 3]
```diff
@@ -782,6 +783,31 @@ func hasSelectorListeners(gw *gatewayapiv1.Gateway) bool {
...
+		}
+
+		return []reconcile.Request{{
+			NamespacedName: types.NamespacedName{
+				Name:      defaultName,
+				Namespace: defaultNs,
+			},
+		}}
+	})
+}
```
Comment (pierDipi) [inline internal/controller/serving/llm/gateway_controller.go]: should we "global resync" here instead of only the default GW?

Comment (bartoszmajsak) [reply internal/controller/serving/llm/gateway_controller.go]: Yeah, this is too narrow. Embrassingly missed it :\

**Comment by bartoszmajsak** [comment]
For the new job I think I had a local kind with some extra setup - I will fix it for the authz test coverage. It worked on my machine! :D https://gist.github.com/bartoszmajsak/9e101e4c4b4a2e1d0f5e5a03d3e9ce86

File: internal/controller/resources/template/PUBLISHER-PATH-AUTH.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+3. **Gateway controller**: read `modelBasedRoutingHeaderName` from `inferenceservice-config`
+   ConfigMap via `GetModelRoutingHeader()` (lowercased, validated as HTTP token). Watch the
+   ConfigMap's `ingress` key for changes.
+4. **Constant**: `DefaultModelRoutingHeader = "x-gateway-model-name"`.
+
+### kserve midstream overlay
+
+`config/overlays/odh/rbac/user-cluster-roles.yaml` - add aggregated `kserve-model-access`
+ClusterRole granting `serve` on `serving.opendatahub.io/models` so existing viewer/editor
+users retain access when traffic moves to BBR/publisher routing.
```
Comment (pierDipi) [inline internal/controller/resources/template/PUBLISHER-PATH-AUTH.md]: do you have a PR already to not forget ?

File: internal/controller/utils/utils.go [hunk 5]
```diff
@@ -325,6 +325,46 @@ func GetDefaultGatewayRef(ctx context.Context, cli client.Client) (namespace, na
...
+				}
+				log.FromContext(ctx).V(0).Info("invalid modelBasedRoutingHeaderName, using default", "value", header)
+			}
+		}
+	}
+
+	return constants.DefaultModelRoutingHeader
+}
+
+func isValidHTTPHeaderName(name string) bool {
```
Comment (pierDipi) [inline internal/controller/utils/utils.go]: should we add some tests for this since it seems kinda security critical? (maybe through the public function caller), perhaps this validation is better placed in the inferenceservice-config webhook validation so that it's guaranteed to not be weird

File: internal/controller/resources/template/PUBLISHER-PATH-AUTH.md (hunk 4 above)
Comment (bartoszmajsak) [reply internal/controller/resources/template/PUBLISHER-PATH-AUTH.md]: About to open

File: internal/controller/test/e2e/publisher_path_authpolicy_test.go [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+func TestBBRPathWithoutHeaderAuthnOnly(t *testing.T) {
+	t.Parallel()
+	f := setupPublisherFixture(t)
+
+	// /v1/chat/completions without model header: no authorization rule fires, authn-only.
+	resp, _ := batchEnv.gatewayGet(t, "/v1/chat/completions", f.noAccessToken, nil)
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("expected 200 (authn-only), got %d", resp.StatusCode)
+	}
+}
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/controller/test/e2e/publisher_path_authpolicy_test.go]: Isn't it stale? TestBBRPathWithoutHeaderAuthnOnly

File: internal/controller/utils/utils.go (hunk 5 above)
Comment (bartoszmajsak) [reply internal/controller/utils/utils.go]: Reworked to use https://pkg.go.dev/golang.org/x/net/http/httpguts#ValidHeaderFieldName

File: .github/workflows/test-authpolicy-e2e.yml [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+          kind: Cluster
+          apiVersion: kind.x-k8s.io/v1alpha4
+          nodes:
+            - role: control-plane
+            - role: worker
+          EOF
+
+      - name: Install MetalLB
+        run: |
+          kubectl apply -f "https://raw.githubusercontent.com/metallb/metallb/${METALLB_VERSION}/config/manifests/metallb-native.yaml"
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply .github/workflows/test-authpolicy-e2e.yml:68]: aren't release artifacts immutable?

File: .github/workflows/test-authpolicy-e2e.yml [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+      - 'internal/controller/resources/template/authpolicy_*.yaml'
+      - 'internal/controller/resources/authpolicy.go'
+      - 'internal/controller/resources/options.go'
+      - 'internal/controller/test/e2e/**'
+      - 'internal/controller/serving/llm/gateway_controller.go'
+      - 'internal/controller/utils/utils.go'
+      - 'internal/controller/constants/constants.go'
+      - '.github/workflows/test-authpolicy-e2e.yml'
+      - 'hack/setup-authpolicy-e2e.sh'
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply .github/workflows/test-authpolicy-e2e.yml:16]: This lets you manually re-run the workflow for debugging without pushing a commit. Removing it is not helpful.

File: .github/workflows/test-authpolicy-e2e.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply .github/workflows/test-authpolicy-e2e.yml:68]: Follow-up?

File: internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml [hunk 9]
```diff
@@ -19,15 +30,100 @@ spec:
...
+              value: post-delegate
+        priority: 1
+      # Deny requests where routing identity (header) and authorization identity
+      # (path) would diverge. A per-participant path with a valid model header
+      # would be routed by the header (tenant-B) but authorized by the path
+      # (tenant-A). patternMatching with predicate "false" is an immediate local
+      # deny - no API server call.
+      deny-misrouted-model-header:
+        when:
+          - predicate: "!request.path.startsWith('/v1/')"
```
Comment (pierDipi) [inline internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml:106]: should this be only "is batch" type of check?

File: internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml [hunk 10]
```diff
@@ -19,15 +30,100 @@ spec:
...
+              && request.headers['{{.ModelRoutingHeader}}'].matches('^publishers/[^/]+/models/.+$')
+        patternMatching:
+          patterns:
+            - predicate: "false"
+        priority: 0
       inference-access:
         when:
-          - predicate: "!(request.path == '/v1/files' || request.path.startsWith('/v1/files/') || request.path == '/v1/batches' || request.path.startsWith('/v1/batches/'))"
+          - patternRef: has-ns-and-name
+          - predicate: "!request.path.startsWith('/v1/')"
```
Comment (pierDipi) [inline internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml:119]: and same here?

File: internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml (hunk 9 above)
Comment (pierDipi) [reply internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml:106]: because I'm wondering if we're allowing requests with ModelRoutingHeader + /v1/completions without any authorization

**Review by pierDipi — APPROVED** [review]
we can follow up

/lgtm
/approve

File: internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml (hunk 10 above)
Comment (bartoszmajsak) [reply internal/controller/resources/template/authpolicy_llm_isvc_userdefined.yaml:119]: Reverting it to batch-only exclusion would run garbage SAR checks on all `/v1/` paths. I think this is arguably better, unless we support BBR for this path, which I thought we didn't?

When the header is present, it would be caught in `deny-misrouted-model-header` right now.

For top-level `/v1/models` etc I think a dedicated `AuthPolicy` on its `HTTPRoute` would make sense anyway?
